stl/inc/__msvc_sanitizer_annotate_container.hpp

```diff
@@ -15,7 +15,40 @@ _STL_DISABLE_CLANG_WARNINGS
 #pragma push_macro("new")
 #undef new
 
-#if !defined(_M_CEE_PURE) && !(defined(_DISABLE_STRING_ANNOTATION) && defined(_DISABLE_VECTOR_ANNOTATION))
+// The following macros change the behavior of this file:
+//   - _DISABLE_STL_ANNOTATION: Disable ASan annotations in the standard library
+//     (this will be auto-defined on unsupported platforms)
+//     + _DISABLE_STRING_ANNOTATION: same, but for only `basic_string`
+//     + _DISABLE_VECTOR_ANNOTATION: same, but for only `vector`
+//   - _ENABLE_STL_ANNOTATION_ON_UNSUPPORTED_PLATFORMS: Don't auto-disable ASan annotations
+//   - _ANNOTATE_STL: Even when ASan annotations are disabled, insert the code for annotating into the STL anyways;
+//     this is useful when building static libraries which may be linked against both ASan and non-ASan binaries.
+//     + _ANNOTATE_STRING: same, but only for `basic_string`
+//     + _ANNOTATE_VECTOR: same, but only for `vector`
+
+#if !defined(_DISABLE_STL_ANNOTATION) && !defined(_ENABLE_STL_ANNOTATION_ON_UNSUPPORTED_PLATFORMS)
+
+#if defined(_M_ARM64EC) || defined(_M_ARM64) || defined(_M_ARM) || defined(_M_CEE_PURE)
+#define _DISABLE_STL_ANNOTATION
+#endif // ^^^ unsupported platform
+
+#endif // ^^^ !defined(_DISABLE_STL_ANNOTATION) && !defined(_ENABLE_STL_ANNOTATION_ON_UNSUPPORTED_PLATFORM)
+
+#ifdef _DISABLE_STL_ANNOTATION
+
+#ifdef _ENABLE_STL_ANNOTATION_ON_UNSUPPORTED_PLATFORMS
+#error _ENABLE_STL_ANNOTATION_ON_UNSUPPORTED_PLATFORMS and _DISABLE_STL_ANNOTATION are mutually exclusive
+#endif // ^^^ defined(_ENABLE_STL_ANNOTATION_ON_UNSUPPORTED_PLATFORMS)
+
+#define _DISABLE_STRING_ANNOTATION
+#define _DISABLE_VECTOR_ANNOTATION
+
+#endif // ^^^ defined(_DISABLE_STL_ANNOTATION)
+
+#ifdef _ANNOTATE_STL
+#define _ANNOTATE_STRING
+#define _ANNOTATE_VECTOR
+#endif // ^^^ defined(_ANNOTATE_STL)

```

Comment:
This will warn if the user has accumulated both `/D_ANNOTATE_STL` and either `/D_ANNOTATE_STRING` or `/D_ANNOTATE_VECTOR` in their build system:

```
C:\Temp>type cats.cpp
#ifdef ANNOTATE_STL
#define ANNOTATE_VECTOR
#endif

C:\Temp>cl /EHsc /nologo /W4 /c cats.cpp
cats.cpp

C:\Temp>cl /EHsc /nologo /W4 /c /DANNOTATE_STL /DANNOTATE_VECTOR cats.cpp
cats.cpp
cats.cpp(2): warning C4005: 'ANNOTATE_VECTOR': macro redefinition
cats.cpp(2): note: 'ANNOTATE_VECTOR' previously declared on the command line
```

It seems harmless to enable both the coarse-grained and fine-grained options here, so I recommend wrapping each of these in `#ifndef`s.